### ISSUE TITLE
User with dollar '$' character in password cannot login in UI for File Based auth (/etc/st2/htpasswd)

### ISSUE THREAD

**Issue by leopoul** [StackStorm/st2#3216]
Version: st2 2.1.1
User with '$' character in password fails to login in UI when File Based auth is used.
eg. 
username: user
password: pas$word
gets " Invalid or missing credentials" 

Issue does not occur when authenticating locally in ST2 instance:
$ st2 auth user -p pas$word

> +----------+----------------------------------+
> | Property | Value                            |
> +----------+----------------------------------+
> | user     | user                             |
> | token    | 968a52ac3ad34858a9285b49b66a43e0 |
> | expiry   | 2017-02-14T13:56:38.629161Z      |

Changing password to password:
echo "password" | sudo htpasswd -i /etc/st2/htpasswd user

logs in the user in UI.

Maybe other special characters introduce the issue.

Best,
Leo

**Comment by Kami** [comment]
Thanks for reporting this - we will have a look.

It seems like an issue in WebUI - https://github.com/stackstorm/st2web.

**Comment by LindsayHill** [comment]
I'm wondering if it's to do with escaping special chars in the shell? E.g. If you run `st2 auth user -p pas$word`, then the shell is going to expand $word. More likely it's unset, so the command will look like `st2 auth user -p pas`?

**Comment by LindsayHill** [comment]
I just tested this with these steps:
```
[vagrant@st2vagrant ~]$ echo 'pas$word' | sudo htpasswd -i /etc/st2/htpasswd newuser
Adding password for user newuser
[vagrant@st2vagrant ~]$ st2 auth newuser -p 'pas$word'
+----------+----------------------------------+
| Property | Value                            |
+----------+----------------------------------+
| user     | newuser                          |
| token    | 2492ec6310814f4d9cde32849565996f |
| expiry   | 2017-02-14T15:54:57.695863Z      |
+----------+----------------------------------+
```

I then successfully logged into the Web UI using newuser/pas$word

@leopoul, when you first created that user with the password 'pas$word', what command did you run?

Did you use `echo "pas$word" | sudo htpasswd -i /etc/st2/htpasswd user`, or did you use `echo 'pas$word' | sudo htpasswd -i /etc/st2/htpasswd user`? Or no quotes at all around `pas$word`?

**Comment by lakshmi-kannan** [comment]
https://github.com/StackStorm/st2/issues/3134 Related.

**Comment by LindsayHill** [comment]
See #3134 - that addresses the use-case where `:` is used. The issue here with `$` is almost certainly related to shell escaping

**Comment by LindsayHill** [comment]
Addressed with #3271
